docs/design/features/NonGC-Heap.md

```diff
@@ -0,0 +1,125 @@
+* **Immortal Objects Only:** Only immortal objects should be allocated on NonGC heap.
+* **Avoid unloadble contexts:** NonGC heap should never be used for objects belonging to unloadable contexts. Doing so could lead to potential memory leaks. As a consequence, all string literals associated with **unloadable** Assembly Load Contexts (ALCs) won't be placed in the NonGC Heap.
+* **No References to GC Heap's Objects:** NonGC heap **must** never contain references to GC heaps' objects. It may contain references to other NonGC objects.
+* The current design of CoreCLR's `FrozenObjectHeapManager` has certain design limitations such as:
+  * Large objects are not supported.
+  * Objects requiring unusual data alignment are not supported.
+
+Because of these restrictions, users **must never** assume that certain kinds of objects can be accessed without pinning.
+
+## What typically can be found in an NonGC heap?
+The NonGC Heap is suitable for a specific set of objects that adhere to the previously described rules. Here's a list showcasing the types of objects that are commonly found in the NonGC Heap for CoreCLR and NativeAOT runtimes:
+* String literal objects
+* RuntimeType objects (CoreCLR-only as of the time of writing), which means for:
+```cs
+Type Test() => typeof(int);
+```
+we used to emit:
+```asm
+; Method Test():System.Type:this
+       sub      rsp, 40
+       mov      rcx, 0x7FFC892DE5D8
+       call     CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE
+       nop      
+       add      rsp, 40
+       ret      
+; Total bytes of code: 25
+
+```
+but with the help of NonGC heap, it's now:
+```asm
+; Method Test():System.Type:this
+       mov      rax, 0x2C1802055E8 ; 'System.Int32' RuntimeType object
+       ret      
+; Total bytes of code: 11
+```
+* Boxed statics:  All static fields of non-primitive value types wrap values in objects to store them on the heap. With NonGC Heap, we can avoid indirect access and reach the field's value directly, further enhancing the efficiency of these operations.
+* Certain kinds of `static readonly` fields which can be pre-initialized on NativeAOT or recognized by CoreCLR e.g., `Array<>.Empty`.
+
+## Diagnostics and visible changes in .NET 8.0
+The most visible change that now `GC.GetGeneration(object)` API returns `int.MaxValue` for such objects. This change serves to emphasize that these objects are not managed by the garbage collector. Similiarly, corresponding `ICorProfiler` APIs may return the same value as a generation. A full inventory of new APIs and alterations related to diagnostic support for the NonGC Heap can be found in the ["Diagnostic support for NonGC heap"](https://github.com/dotnet/runtime/issues/75836) issue. An important detail to keep in mind that these object should never be considered as "dead" if noone references them since it's possible they're actually kept by unmanaged data structures in VM itself.
```

Comment:
```suggestion
The most visible change that now `GC.GetGeneration(object)` API returns `int.MaxValue` for such objects. This change serves to emphasize that these objects are not managed by the garbage collector. Similiarly, corresponding `ICorProfiler` APIs may return the same value as a generation. A full inventory of new APIs and alterations related to diagnostic support for the NonGC Heap can be found in the ["Diagnostic support for NonGC heap"](https://github.com/dotnet/runtime/issues/75836) issue. An important detail to keep in mind is that these objects should never be considered as "dead" if no one references them since it's possible that they're actually kept alive by unmanaged data structures in VM itself.
```
?

Reply:
Thanks!